### ISSUE TITLE
[File Tree] Needs filter options

### ISSUE THREAD

**Issue by taiyab** [sourcegraph/sourcegraph#58427]
The LHS (left-hand side) file tree view needs the ability to filter through files/dirs by extension or programming language (this one is a bit weird).

<img width="285" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/5337876/524416a1-20f9-4254-9b01-1bb0cf13e903">

It should:
- Directly interact and match the state of the headline query (i.e. let's say I've defined `file:*.tsx`, it should only highlight Typescript files in the LHS, and that matching filter query should be input in a filter bar above the file tree to match.

@rrhyne This should also help inform/solve your conundrum around how the query impacts the file tree view on say the repo page. I feel like applying file-level filters should never completely hide files that don't match the query (specifically in the file tree), instead, it should filter the list by query + alphabetical, then appending things that didn't match the file-level filter to the end of the list, and put them in a dull state.

Basically, it reinforces the idea I've started to play with which is the UI should reflect the filters you've applied in headline query throughout, then you're free to modify them at a "local" level to modify the query in a simpler way when tweaking.

## Things to figure out
- How `lang: x` might impact filtering files by extension
- How do deal with languages that are broken up across multiple filetypes (e.g. Obj-C, `*.h` and `*.m` filetypes, and sure there's even more complex cases than this, and overlap).

**Comment by camdencheek** [comment]
Closing as old since this will also be covered as part of the ongoing redesigns.
